Cython/Compiler/Parsing.py

```diff
@@ -1492,9 +1507,9 @@ def p_testlist_star_expr(s):
 
 # testlist_comp: (test|star_expr) ( comp_for | (',' (test|star_expr))* [','] )
 
-def p_testlist_comp(s):
+def p_testlist_comp(s, is_expression=False):
```

Comment:
I think some of these parse functions could benefit from being split in two (or otherwise refactored a bit), rather than passing them an option. For example, it's not clear to me what `is_expression=False` means for an expression.

At some point, we should also try to come a little closer again to CPython's parser structure. Cython's structure was originally based on the old CPython grammar but has probably diverged over the years. Now that CPython has a new parser overall (generated from [Parser/Python.asdl](https://github.com/python/cpython/blob/main/Parser/Python.asdl) as [Parser/parser.c](https://raw.githubusercontent.com/python/cpython/main/Parser/parser.c)), it's probably worth taking some guidance from that.

Reply:
I consider this a required refactoring since the addition of these flags really degrades the grammar structure of the parser, but given that the parser already needs a bit of a cleanup, the change would be non-trivial, and this PR isn't a good place to do that, I can accept leaving it as it is now and moving the refactoring to another PR.